### PR TITLE
GeoTile Aggregation

### LINKED ISSUE #449: _(geo)aggregate :  Add geotile aggregation type
As a user of ARLAS-server `_aggregate` service

I want to be able to aggregate geo-points using a `map tile` grid *({zoom}/{x}/{y})* with a large precision range (larger than the geohash grid aggregation precision range)

in order to smooth the precision changes while zooming in (in a map client for example)

According to the documentation 
https://www.elastic.co/guide/en/elasticsearch/reference/7.11/search-aggregations-bucket-geotilegrid-aggregation.html

Technically this means :
- [x] new aggregation type **geotile**
  - datehistogram
  - histogram
  - term
  - geohash
  - **geotile**

- [x] new endpoint `@Path("{collection}/_geoaggregate/{z}/{x}/{y}")
   - so we need to transform {z}/{x}/{y} EPSG 3857 tiles to a bounding box as top_left and bottom_right. (same behaviour of the endpoint `@Path("{collection}/_geoaggregate/{geohash}")`)
- [x] For this new agg type, we need to pass a precision in the `agg` query param, in the `interval` attribute (default value `z+3` if not specified)
   - this precision corresponds to zoom level of the aggregation, that should be larger than or equal to {z} in the path param
   - the maximum difference between the precision and {z} in the path param is **6**
   - If we want a difference of 7 we need to split the  bbox resulting from the {z}/{x}/{y} tile into 4 similar bboxes the same way we did #656 
- [x] How to handle returned geometries by the ES aggregation ?
  - The response of ES is 
```
   "aggregations": {
    "large-grid": {
      "buckets": [
        {
          "key": "8/131/84",
          "doc_count": 3
        },
        {
          "key": "8/129/88",
          "doc_count": 2
        },
        {
          "key": "8/131/85",
          "doc_count": 1
        }
      ]
    }
    }
```
We need to transform the `"key"` to a geometry. The geometry to return is to be specified in `aggregated_geometries` subparam of the `agg` param.

For now `aggregated_geometries` have four possible values
- geohash
- geohash_center
- centroid
- bbox

- [x] First thing we need to do, is replace 
`geohash` => `tile`
`geohash_center` => `tile_center`
in order to be generic to the `**geotile**` and `**geohash**`

- [x] Update documentation
- [ ] Integration tests

### PR DESCRIPTION
**Pull request by alainbodiguel**
closes #449

### CODE REVIEW

File: arlas-core/src/main/java/io/arlas/server/utils/CheckParams.java [hunk 1]
```diff
@@ -66,15 +66,17 @@
     private static final String INVALID_ON_VALUE = "Invalid 'on-' value : ";
     private static final String BAD_COLLECT_FCT_COLLECT_FIELD_NUMBERS = "'collect_field' and 'collect_fct' occurrences should be even.";
     private static final String DATE_NOW = "now";
-    private static final String AGGREGATED_GEOMETRY_NOT_SUPPORTED = "'geohash' & 'geohash_center' are only supported for geohash aggregation type.";
+    private static final String AGGREGATED_GEOMETRY_NOT_SUPPORTED = "'tile' & 'tile_center' are only supported for geohash nd geotile aggregation type.";
```
Comment (MohamedHamouGisaia) [inline arlas-core/src/main/java/io/arlas/server/utils/CheckParams.java:69]: nd => and

File: arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java [hunk 2]
```diff
@@ -33,30 +31,56 @@
...
         pwithin = "geo_params.centroid:within:-5,0,0,5";
-        handleGeohashTileDisjointFromPwithin(geohashTilePwithinGet("geohash:geo_params.centroid:interval-3", Arrays.asList(pwithin), "yn"));
+        handleGeohashTileDisjointFromPwithin(tilePwithinGet("geohash:geo_params.centroid:interval-3", Arrays.asList(pwithin), "yn"));
+
+    }
+
+    @Test
+    public void testGeotile() throws Exception {
+        //GEOTILE
+        // precision = geohashLength  ==> we should have one feature maximum
```
Comment (MohamedHamouGisaia) [inline arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java:64]: precision = {z}

File: arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java [hunk 3]
```diff
@@ -33,30 +31,56 @@
...
+        handleGeotileGreaterThanPrecision(tileGet("geotile:geo_params.centroid:interval-5", "5/4/3"), 1, "5/4/3");
+
+        // precision > geohashLength  ==> we could have more than one feature
```
Comment (MohamedHamouGisaia) [inline arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java:67]: precision > {z}

File: arlas-core/src/main/java/io/arlas/server/utils/CheckParams.java (hunk 1 above)
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/utils/CheckParams.java]: fixed

File: arlas-core/src/main/java/io/arlas/server/impl/elastic/core/FluidSearch.java [hunk 4]
```diff
@@ -664,6 +673,20 @@ private GeoGridAggregationBuilder buildGeohashAggregation(Aggregation aggregatio
         return geoHashAggregationBuilder;
     }
 
+    // construct and returns the geohash aggregationModel builder
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/impl/elastic/core/FluidSearch.java]: fixed

File: arlas-core/src/main/java/io/arlas/server/app/Documentation.java [hunk 5]
```diff
@@ -213,6 +215,7 @@
     public static final String GEOAGGREGATION_OPERATION = "Aggregate the elements in the collection(s) as features, given the filters and the aggregation parameters.";
     public static final String SHAPEAGGREGATION_OPERATION = "Aggregate the elements in the collection(s) as features, given the filters and the aggregation parameters, and returns a shapefile of it.";
     public static final String GEOHASH_GEOAGGREGATION_OPERATION = "Aggregate the elements in the collection(s) and localized in the given geohash as features, given the filters and the aggregation parameters.";
+    public static final String GEOTILE_GEOAGGREGATION_OPERATION = "Aggregate the elements in the collection(s) and localized in the given geohash as features, given the filters and the aggregation parameters.";
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/app/Documentation.java]: fixed

File: arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java (hunk 2 above)
Comment (alainbodiguel) [reply arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java]: fixed

File: arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java (hunk 3 above)
Comment (alainbodiguel) [reply arlas-tests/src/test/java/io/arlas/server/rest/explore/AbstractGeohashTiledTest.java]: fixed

**Pull request merged**